README.md
# sqlrat

SQL template library for Clojure, ClojureScript and ClojureCLR

It helps generate well formed SQL using templates and entities. While one can write hand-crafted templates to generate complex SQL, entities are useful to easily generate mundane [CRUD](http://en.wikipedia.org/wiki/Create,_read,_update_and_delete) SQL via auto-generated templates. This library does not execute SQL; it only outputs appropriate SQL with placeholders for JDBC, ODBC, ADO.NET and Node.js libraries.

## Usage

### Leiningen/Maven dependency

Not on Clojars yet. Run `lein pkg install` to install locally and include `[sqlrat "0.3.0-SNAPSHOT"]` in project dependencies.

### Namespaces

```clojure
(ns example
  (:require [sqlrat.template :as t]
            [sqlrat.entity :as e])
;; Require macros for Clojure and ClojureCLR
  (:require [sqlrat.macros :refer (deftemplate defentity)])
;; Require macros for ClojureScript
  (:require-macros '[sqlrat.macros :refer (deftemplate defentity)])
;; other stuff
  ..)
```

### Template

Templates are fundamental building blocks in _Sqlrat_ supporting identifier and value placeholders.

#### Keyword arguments as value placeholders

All of the following define a var `s` to the **same template** (with keyword args `:dept-id` and `:active`):

```clojure
(def s (t/make-template ["SELECT * FROM emp WHERE dept_id =" :dept-id "AND active =" :active]))
(def s (t/parse-template "SELECT * FROM emp WHERE dept_id = :dept-id AND active = :active"))
(def s #sqlrat/template ["SELECT * FROM emp WHERE dept_id =" :dept-id "AND active =" :active])
(def s #sqlrat/template "SELECT * FROM emp WHERE dept_id = :dept-id AND active = :active")
(deftemplate s ["SELECT * FROM emp WHERE dept_id =" :dept-id "AND active =" :active])
(deftemplate s "SELECT * FROM emp WHERE dept_id = :dept-id AND active = :active")
```

Templates must be realized in order to generate the SQL and arguments:

```clojure
(t/realize s {:dept-id 10 :active true})
;;=> ("SELECT * FROM emp WHERE dept_id = ? AND active = ?" 10 true)
```

##### Multi-value keyword arguments

Keywords can also act as multi-value placeholders, for example the following:

```clojure
(def t #sqlrat/template "SELECT * FROM emp WHERE id IN (:ids)")
(t/realize t {:ids [1 2 3]})
;;=> ("SELECT * FROM emp WHERE id IN ( ?, ?, ? )" 1 2 3)
```

#### Symbol arguments as identifier placeholders

Templates also support symbol arguments as identifier placeholders:

```clojure
(def t (t/make-template ["SELECT * FROM emp WHERE" 'dept-col "=" :dept-val]))
(t/realize t {'dept-col "dept_id" :dept-val 20})
;;=> ("SELECT * FROM emp WHERE dept_id = ?" 20)
```

Multi-values symbol arguments are joined using a default identifier delimiter.

```clojure
(t/realize t {'dept-col ["emp" "dept_id"] :dept-val 20})
;;=> ("SELECT * FROM emp WHERE emp.dept_id = ?" 20)
```

##### Identifier decoration

Should you want to quote the database identifiers, specify the `:subst` arity-1 (identifier-name) fn in options:

```clojure
(t/realize t {'dept-col ["emp" "dept_id"] :dept-val 20} {:subst #(str "`" % "`")})
;;=> ("SELECT * FROM emp WHERE `emp`.`dept_id` = ?" 20)
```

#### Value placeholder decoration

When using PostgreSQL, you may have to specially decorate the `?` placeholders for integers on the JVM:

```clojure
(t/realize t {'dept-col "dept_id" :dept-val 20} {:place (t/pgsql-wrap-place)})
;;=> ("SELECT * FROM emp WHERE dept_id = ?::integer" 20)
```

Unless you are on the JVM, you may not want the `?` placeholder in the rendered SQL - use a custom arity-2 (index and column-name) `:place` function

```clojure
;; for some Node.js database libraries
(t/realize t {'dept-col "dept_id" :dept-val 20} {:place t/dollar-place})
;;=> ("SELECT * FROM emp WHERE dept_id = $1" 20)

;; for ADO.NET SqlCommand API
(t/realize t {'dept-col "dept_id" :dept-val 20} {:place t/at-param-place})
;;=> ("SELECT * FROM emp WHERE dept_id = @param1" 20)
```

### Entity

Writing templates for mundane CRUD jobs is tiring, which is where entities help. An entity represents a database entity metadata and can be defines as follows:

```clojure
(defn ts-now [] (java.sql.Timestamp. (System/currentTimeMillis)))

(defentity BlogPost {:table "post" :id [:post-id]}
  :post-id {:insert? false}  ; column won't be included in INSERT statement
  :content {}
  :active  {:default true}   ; default value is supplied in INSERT statement
  :created {:default ts-now} ; default value is obtained by executing fn
  :updated {:default ts-now})
```

#### CRUD operations

A CRUD operation on an entity creates a template that you can realize just like a hand-crafted template.

##### INSERT

```clojure
(def ip (e/insert-row BlogPost))  ; returns a template
(t/realize ip {:content "Foo bar"})
;;=> ("INSERT INTO post ( content , active , created , updated ) VALUES ( ? , ? , ? , ? )"
;;..  "Foo bar" true #inst "2013-09-09T12:40:52.765000000-00:00"
;;..  #inst "2013-09-09T12:40:52.765000000-00:00")
```

##### SELECT

```clojure
(def sp (e/find-by-id BlogPost))
(t/realize sp {:post-id 10})
;;=> ("SELECT post_id , content , active , created , updated FROM post WHERE post_id = ?" 10)
```

See `sqlrat.entity/find-all` and `sqlrat.entity/find-where` for other options.

##### UPDATE

```clojure
(def up (e/update-by-id BlogPost (e/cols [:content])))
(t/realize up {:post-id 10 :content "Foo bar baz"})
;;=> ("UPDATE post SET content = ? WHERE post_id = ?" "Foo bar baz" 10)
```

**Compare-and-swap (optimistic locking)** operations with UPDATE can be done with renamed columns:

```clojure
(def sp (e/update-where BlogPost [:post-id :created]
                        (e/cols {:content :content :created :new-created})))
(t/realize sp {:post-id 10 :content "Foo bar baz quux" :created 9})
;;=> ("UPDATE post SET content = ? , created = ? WHERE post_id = ? AND created = ?"
;;..  "Foo bar baz quux" #inst "2013-09-09T15:13:45.112000000-00:00" 10 9)
```

See `sqlrat.entity/update-all` and `sqlrat.entity/update-where` for other options.

##### DELETE

```clojure
(def dp (e/delete-by-id BlogPost))
(t/realize dp {:post-id 10})
;;=> ("DELETE FROM post WHERE post_id = ?" 10)
```

See `sqlrat.entity/delete-all` and `sqlrat.entity/delete-where` for other options.


### Entity relations

TODO


## Development

### Running tests

```bash
lein 1.4 test  # test with Clojure 1.4
lein dev test  # test with Clojure 1.5 and ClojureScript
lein clr test  # test with ClojureCLR (env var CLJCLR14_40 should be defined)
```

### Packaging a JAR

```bash
lein pkg jar      # includes cljx, clj and cljs files into the JAR
lein pkg install  # includes cljx, clj and cljs files into the JAR
```

## License

Copyright © 2013 Shantanu Kumar

Distributed under the Eclipse Public License, the same as Clojure.


CHANGES.md
# TODO and Changes


## TODO

* [TODO] Join-based SQL statements (avoid N+1 operations)
   * [TODO] find-rels - given A, [D col1 col2] infer JOIN A..B..C..D (many to many)
   * [TODO] insert-rels - maybe "INSERT SELECT"
   * [TODO] update-rels
   * [TODO] delete-rels


## Changelog

### Version 0.3.0 / 2013-Sep-??

* SQL template
   * Symbols are database identifiers, substituted by values
   * Keywords are value placeholders, replaced by `?`
   * Substitution fn for symbols - e.g. `emp` can become `'emp'`
   * Placeholder fn for keywords - e.g. `?` can become `?::integer` for ints
   * Partial template realization (symbols only)
   * Parser - e.g. turn `"WHERE code = :code"` into `["WHERE code =" :code]`
* Entity
   * Entity metadata (optional keys: `:table` `:id`)
   * Column metadata (optional keys: `:colname` `:insert?` `:default`)
   * CRUD functions (by columns and ID, using symbol-substituted identifiers)
   * Optional map args - distinct, cols, table, where-cols, order-by
* Entity relation
   * Relation mapping (adjacency list) definition
   * Value induction from related entity instances



src/sqlrat/macros.clj
(ns sqlrat.macros
  (:require [sqlrat.template :as t]
            [sqlrat.entity :as e]
            [sqlrat.relation :as rel]))


(defmacro deftemplate
  [sym tokens]
  `(def ~sym (t/make-or-parse-template ~tokens)))


(defmacro defentity
  [sym entity-meta col-kw col-meta & more]
  {:pre [(symbol? sym)
         (map? entity-meta)
         (keyword? col-kw)
         (map? col-meta)
         (even? (count more))
         (every? keyword? (map first (partition 2 more)))
         (every? map? (map second (partition 2 more)))]}
  (let [e-name (name sym)]
    `(def ~sym (e/make-entity ~e-name ~entity-meta
                              (array-map ~col-kw ~col-meta ~@more)))))


(defmacro defrelmap
  [sym & tokens] {:pre [(symbol? sym)
                        (even? (count tokens))]}
  `(def ~sym (rel/make-relmap (array-map ~@tokens))))


(comment
  (defentity Employee {:table "emp" :id [:emp-id]}
    :emp-id      {:insert? false}
    :emp-code    {}
    :first-name  {:colname "f_name"}
    :dept-code   {}
    :created-at  {:default #(java.sql.Timestamp. (System/currentTimeMillis))})
  (defentity Department {:table "dept" :id [:dept-id]}
    :dept-id   {:insert? false}
    :dept-code {}
    :dept-head {})
  (defrelmap rels
    [Employee :dept-code] [Department :dept-code])
  (insert-row Employee)
  (find-by-id Employee)
  (find-by-cols Employee :emp-code)
  (update-by-id Employee)
  (delete-by-id Employee))


project.clj
(defproject sqlrat "0.3.0-SNAPSHOT"
  :description "SQL template library for Clojure, ClojureScript and ClojureCLR"
  :url "https://github.com/kumarshantanu/sqlrat"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :plugins [[com.keminglabs/cljx "0.3.0"]
            [lein-clr "0.2.0"]]
  :clr {:cmd-templates {:clj-url "http://sourceforge.net/projects/clojureclr/files/clojure-clr-1.4.1-Debug-4.0.zip/download"
                        :clj-zip "clojure-clr-1.4.1-Debug-4.0.zip"
                        :clj-dep [#_"mono" ["target/clr/clj/Debug 4.0" %1]]
                        :clj-exe [#_"mono" [CLJCLR14_40 %1]]
                        :wget    ["wget" "--no-check-certificate" "--no-clobber" "-O" %1 %2]
                        :unzip   ["unzip" "-d" %1 %2]}
        :deps-cmds [;[:wget  :clj-zip :clj-url]
                    ;[:unzip "../clj" :clj-zip]
                    ]
        :main-cmd    [:clj-exe "Clojure.Main.exe"]
        :compile-cmd [:clj-exe "Clojure.Compile.exe"]}
  :cljx {:builds [{:source-paths ["src"]  :output-path "target/generated/clj"       :rules :clj}
                  {:source-paths ["test"] :output-path "target/generated/test-clj"  :rules :clj}
                  {:source-paths ["src"]  :output-path "target/generated/cljs"      :rules :cljs}
                  {:source-paths ["test"] :output-path "target/generated/test-cljs" :rules :cljs}]}
  :source-paths ["src"  "target/generated/clj" "resources"]
  :test-paths   ["test" "target/generated/test-clj"]
  :profiles {:jst {:dependencies [[org.clojure/clojurescript #_"0.0-1847" #_"0.0-1853" #_"0.0-1859" "0.0-1877"]
                                  [com.cemerick/clojurescript.test "0.0.4"]]
                   :plugins [[lein-cljsbuild "0.3.3"]]
                   :hooks [leiningen.cljsbuild]
                   :cljsbuild {:builds [{:source-paths ["target/generated/cljs" "target/generated/test-cljs"]
                                         :compiler {:output-to "target/cljs/testable.js"
                                                    :optimizations :whitespace
                                                    :pretty-print true}}]
                               :test-commands {"unit-tests" ["phantomjs" "runners/phantomjs.js" "target/cljs/testable.js"]}}}
             :1.4 {:dependencies [[org.clojure/clojure "1.4.0"]]}
             :1.5 {:dependencies [[org.clojure/clojure "1.5.1"]]}
             :pkg {:source-paths ["target/generated/cljs"]}
             }
  :aliases {"1.4" ["with-profile" "1.4" "do" "clean," "cljx" "once,"]
            "dev" ["with-profile" "1.5,jst" "do" "clean," "cljx" "once,"]
            "clr" ["do" "clean," "cljx" "once," "clr"]
            "pkg" ["with-profile" "pkg" "do" "clean," "cljx" "once,"]
            })

